nexus/reconfigurator/execution/src/external_networking.rs

```diff
@@ -0,0 +1,1202 @@
+                format!(
+                    "failed to delete service VNIC {nic:?} for {kind} zone {}",
+                    z.id
+                )
+            })?;
+        if deleted_nic {
+            info!(log, "successfully deleted Omicron zone vNIC");
+        } else {
+            debug!(log, "Omicron zone vNIC already deleted");
+        }
+    }
+
+    Ok(())
+}
+
+// Helper function to determine whether a given external IP address is
+// already allocated to a specific service zone.
+async fn is_external_ip_already_allocated(
+    opctx: &OpContext,
+    datastore: &DataStore,
+    zone_kind: ZoneKind,
+    zone_id: OmicronZoneUuid,
+    external_ip: OmicronZoneExternalIp,
+    log: &Logger,
+) -> anyhow::Result<bool> {
+    // localhost is used by many components in the test suite.  We can't use
+    // the normal path because normally a given external IP must only be
+    // used once.  Just treat localhost in the test suite as though it's
+    // already allocated.  We do the same in is_nic_already_allocated().
+    if cfg!(test) && external_ip.ip().is_loopback() {
+        return Ok(true);
+    }
+
+    let allocated_ips = datastore
+        .external_ip_list_service(opctx, zone_id.into_untyped_uuid())
+        .await
+        .with_context(|| {
+            format!("failed to look up external IPs for {zone_kind} {zone_id}")
+        })?;
+
+    // We expect to find either 0 or exactly 1 IP for any given zone. If 0,
+    // we know the IP isn't allocated; if 1, we'll check that it matches
+    // below.
+    let existing_ip = match allocated_ips.as_slice() {
+        [] => {
+            info!(log, "external IP allocation required for zone");
+
+            return Ok(false);
+        }
+        [ip] => ip,
+        _ => {
+            warn!(
+                log, "zone has multiple IPs allocated";
+                "allocated_ips" => ?allocated_ips,
+            );
+            bail!(
+                "zone {zone_id} already has {} IPs allocated (expected 1)",
+                allocated_ips.len()
+            );
+        }
+    };
+
+    // We expect this to always succeed; a failure here means we've stored
+    // an Omicron zone IP in the database that can't be converted back to an
+    // Omicron zone IP!
+    let existing_ip = match OmicronZoneExternalIp::try_from(existing_ip) {
+        Ok(existing_ip) => existing_ip,
+        Err(err) => {
+            error!(log, "invalid IP in database for zone"; &err);
+            bail!(
+                "zone {zone_id} has invalid IP database record: {}",
+                InlineErrorChain::new(&err)
+            );
+        }
+    };
+
+    if existing_ip == external_ip {
+        info!(log, "found already-allocated external IP");
+        Ok(true)
+    } else {
+        warn!(
+            log, "zone has unexpected IP allocated";
+            "allocated_ip" => ?existing_ip,
+        );
+        bail!("zone {zone_id} has a different IP allocated ({existing_ip:?})",);
```

Comment:
> So we are explicitly not allowing the external IP for a zone to be changed once it's assigned -- is that correct?

I think that's correct based on how zone configs work today; if you send a sled-agent a Nexus zone config, then send it the same Nexus zone config with a different IP, that won't work. (I'm actually not sure what it will do - maybe it would change the IP the next time time zone is restarted? Maybe the PUT would fail? But Nexus doesn't support changing its external IP on the fly.)

We could certainly relax this requirement if that changes, although at that point I think we'd need to transactionally deallocate the old IP + allocate the new IP?